CompareRaftMembers/compareFollowerDbs.js

```diff
@@ -0,0 +1,272 @@
+/* eslint-disable no-console */
+
+const async = require('async');
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+
+const { Logger } = require('werelogs');
+const Level = require('level');
+
+const DBListStream = require('./DBListStream');
+const DiffStreamOplogFilter = require('./DiffStreamOplogFilter');
+const KeyValueStreamDiff = require('./KeyValueStreamDiff');
+
+const DEFAULT_PARALLEL_SCANS = 4;
+const DEFAULT_LOG_PROGRESS_INTERVAL = 10;
+
+const USAGE = `
+compareFollowerDbs.js
+
+This tool compares offline Metadata leveldb databases of two different
+Metadata nodes, and outputs the differences to the file path given as
+the DIFF_OUTPUT_FILE environment variable.
+
+In this file, it outputs each key that differs as line-separated JSON
+entries, where each entry can be one of:
+
+- [{ key, value }, null]: this key is present on follower #1 but not
+  on follower #2
+
+- [null, { key, value }]: this key is not present on follower #1 but
+  is present on follower #2
+
+- [{ key, value: "{value1}" }, { key, value: "{value2}" }]: this key
+  has a different value between follower #1 and follower #2: "value1"
+  is the value seen on follower #1 and "value2" the value seen on
+  follower #2.
+
+Usage:
+    node compareFollowerDbs.js
+
+Mandatory environment variables:
+    DATABASES1: space-separated list of databases of follower #1 to
+    compare against follower #2
+    DATABASES2: space-separated list of databases of follower #2 to
+    compare against follower #1
+    DIFF_OUTPUT_FILE: file path where diff output will be stored
+
+Optional environment variables:
+    PARALLEL_SCANS: number of databases to scan in parallel (default ${DEFAULT_PARALLEL_SCANS})
+    EXCLUDE_FROM_CSEQS: mapping of raft sessions to filter on, where
+        keys are raft session IDs and values are the cseq value for that
+        raft session. Filtering will be based on all oplog records more
+        recent than the given "cseq" for the raft session. Input diff
+        entries not belonging to one of the declared raft sessions are
+        discarded from the output. The value must be in the following JSON
+        format:
+            {"rsId":cseq[,"rsId":cseq...]}
+        Example:
+            {"1":1234,"4":4567,"6":6789}
+        This configuration would cause diff entries which bucket/key
+        appear in either of the following to be discarded from the output:
+        - oplog of raft session 1 after cseq=1234
+        - or oplog of raft session 4 after cseq=4567
+        - or oplog of raft session 6 after cseq=6789
+        - or any other raft session's oplog at any cseq
+    BUCKETD_HOSTPORT: ip:port of bucketd endpoint, needed when EXCLUDE_FROM_CSEQS is set
+`;
+
+for (const envVar of [
```

Comment:
we should move these env var existence checks into arsenal at some point

Reply:
I wouldn't be fond of executing thinks like `process.exit()` in a library function, as I think it's a bad practice to hide control flow changes except in very specific circumstances. Now why not having the rest of the logic in Arsenal, but it's very small and we would still end up doing `if (checkVars(...)) { process.exit(1) } at least so I don't think there would be any real benefit (but some overhead in finding out where the function is etc.). It's still a pattern we have regularly in the various scripts I agree but we don't do much more than logging and exiting in that case.

Reply:
What could make more sense IMHO is using an existing library more often to handle a lot of parameter passing etc. for us and automate the sanity checks. Scuba uses `convict` (https://github.com/scality/scuba/blob/development/1.0/lib/config.ts#L118) which also supports environment variables, maybe we could benefit from it as well in the future (TBH I don't want to tackle this for this internal scan script as I already tested it and I'm lazy :laughing: )

Reply:
covict supports cli flags as well